proposals/Imperative-Shadow-DOM-Distribution-API.md

```diff
@@ -54,36 +76,55 @@ partial interface HTMLSlotElement {
 
 `slot.assign(sequence<Node> nodes)` runs the following steps:
 
-1. Set the slot's *manually-assigned-nodes* to *nodes*.
-2. Run [assign slotables for a tree] with slot’s tree.
+1. Check `slotAssignment` is "manual", throw otherwise.
+2. Validate each node in nodes that its node.parentNode == slot.rootNode.host. Throw if not equal.
+3. Set the slot's `manually-assigned-nodes` to `nodes`.
+4. Run [assign slotables for a tree] with slot’s tree.
 
-step 2 is required because we have to re-calculate [assigned nodes] of every slots in the tree at this timing.
+Step 4 is required because we have to re-calculate [assigned nodes] of every slot in the tree at this timing.
 
-Note: The detail is explained later, however, it would be worth noting that *manually-assigned-nodes* is not used as
-[assigned nodes] as is. You can think that `slot.assign(sequence<Node> nodes)` tell the engine *candidate nodes* from where [assigned nodes] are constructed.
+Note: The detail is explained later, however, it would be worth noting that `manually-assigned-nodes` is not used as
+[assigned nodes] as is. You can think of `slot.assign(sequence<Node> nodes)` as telling the engine a set of "candidate nodes" from which [assigned nodes] are constructed.
 
 [assign slotables for a tree]: https://dom.spec.whatwg.org/#assign-slotables-for-a-tree
 
-# Changes to DOM Standard
+## Changes to DOM Standard
 
-## ShadowRootInit
+### ShadowRootInit
 
 
 ``` webidl
 ShadowRootInit {
    ...
-   (optional) sloting: 'manual'|'auto' //  (if omitted, it is 'auto');
+   (optional) slotAssignment: 'manual'|'auto' //  (if omitted, it is 'auto');
 }
-
 ```
 
-## 4.2.2.3. Finding slots and slotables
+### [concept node insert](https://dom.spec.whatwg.org/#concept-node-insert)
 
-[To find a slot] need to be updated.
-Other steps don't need to be updated from the standard's perspective, I think, thanks to well-factored each steps.
+7. For each node in nodes, in tree order:
 
-[To find a slot] for a given slotable *slotable* and an optional open flag (unset unless stated otherwise), run these steps:
+    1. Adopt node into parent’s node document.
+
+    2. If child is null, then append node to parent’s children.
+
+    3. Otherwise, insert node into parent’s children before child’s index.
+    
+    4. [Update Step] If parent is a shadow host, its shadow's `slotAssignment` is not "manual" and node is a slotable, then assign a slot for node.
+
+    5. [New Step] If parent's root is a shadow root, shadow's `slotAssignment` is "manual", and parent is a slot, throw exception.
```

Comment:
This prevents slots in shadow roots with manual distribution from having fallback content. Is that intended?

Reply:
bicknellr@,

Yes, it's intended, but not sure if the spec wording is correct. Or if it needs to throw an exception vs no-op.

To me, fallback content makes it confusing when working in slot assignment manual mode. Slot assignment should be absolute and final in manual mode.  See comment from TPAC: https://github.com/whatwg/html/issues/3534#issuecomment-537802687

In that case, when developers call slot.assign([]), the slot should be cleared and have no content. But if we allow failback content, then content can still appear inside the slot even thou developer specifically want to clear it out. 

This is my reasoning for update the spec. Let me know if you think I'm overlooking something.

thanks

Reply:
bicknellr@,

After some considering and discussing it with masonfreed@, we decided it's better to preserve the fallback content behavior. So for a shadow root, regardless of whether slotAssignment is manual or auto, the fallback behavior is identical. 

I've since removed this section in the explainer doc. Thanks again for your review.

Reply:
Thanks for taking another look at this. I think that's the best option given that it makes the behavior the same in both cases and that slots' fallback content is very likely controlled by the same entity that would assign nodes.